File: src/wisp/central/api/replay.py
from __future__ import annotations

# THE TIME MACHINE'S one endpoint. It answers "what did this org's fleet look
# like at any moment in the last N days" with an INTERVAL LIST, never a
# per-tick sample: months-deep outage tables exist and a replay that shipped a
# state per device per minute would be a megabyte a day of arithmetic the
# browser can do from three small lists.
#
# The reply is deliberately three kinds of fact, not one:
#   spans   — what the record SAYS (outage rows, the same set
#             analytics.device_reliability reads, so the map replay, the
#             availability strip and the reliability table can never disagree
#             about when a box was down);
#   devices — each device's own recording FLOOR (created_at);
#   blind   — windows in which the probe covering a device was silent.
# The last two exist because `unknown` is a first-class state on the client.
# Without them the reconstruction can only say up/down, and "no outage row
# covers 03:00 last Tuesday" would render as a green pin over four hours
# nobody was watching — the frozen doctrine's exact failure, moved into the
# past.
#
# Epoch SECONDS on the wire (the historian's convention, not the ISO-TEXT one)
# because every consumer does scrub arithmetic on these numbers.

from wisp.central import analytics as central_analytics
from wisp.central.api.common import (org_or_400, q_int_or, reader_or_401,
                                     visible_device_ids)
from wisp.central.history import epoch_s

# The window cap. 90 days is the outage table's useful depth for a REPLAY (the
# reliability strip goes to a year, but nobody scrubs a year minute by minute),
# and it bounds the one unindexed-ish read here.
MAX_DAYS = 90
DEFAULT_DAYS = 7

# An interval: (start, end) in epoch seconds; end None means "still open at
# the end of the window".
Interval = tuple[int, "int | None"]

_OPEN = float("inf")


def _hi(iv: Interval) -> float:
    return _OPEN if iv[1] is None else iv[1]


def stale_intervals(marks: list[dict], since: int, until: int,
                    grace: int) -> dict[str, list[Interval]]:
    # The watchdog's NODE_STALE / NODE_OK transitions folded into per-probe
    # silent windows.
    #
    # A NODE_STALE row is BACK-DATED by the stale threshold, and that is the
    # honest direction rather than the generous one: the watchdog only calls a
    # probe stale BECAUSE nothing has arrived for `central_node_stale_s`, so
    # the silence demonstrably began that long before the row was written.
    # Taking the row's own timestamp would paint three minutes of "up" over a
    # gap we can prove existed.
    out: dict[str, list[Interval]] = {}
    open_at: dict[str, int] = {}
    for m in marks:
        node = m["node_id"]
        try:
            at = epoch_s(m["at"])
        except (ValueError, TypeError):
            continue
        if m["kind"] == "NODE_STALE":
            if node not in open_at:
                open_at[node] = max(since, at - grace)
        elif node in open_at:
            start = open_at.pop(node)
            end = max(start, at)
            if end > since and start < until:
                out.setdefault(node, []).append((max(since, start), end))
    for node, start in open_at.items():
        if start < until:
            out.setdefault(node, []).append((max(since, start), None))
    return out


def intersect(a: list[Interval], b: list[Interval]) -> list[Interval]:
    # Both sides sorted and non-overlapping. Used for a device with NO probe
    # assignment: the pre-assignment default is "every node for this org
    # covers it", so such a device is only unanswerable when EVERY probe was
    # silent at once.
    out: list[Interval] = []
    i = j = 0
    while i < len(a) and j < len(b):
        lo = max(a[i][0], b[j][0])
        hi = min(_hi(a[i]), _hi(b[j]))
        if hi > lo:
            out.append((lo, None if hi == _OPEN else int(hi)))
        if _hi(a[i]) < _hi(b[j]):
            i += 1
        else:
            j += 1
    return out


def device_blind(floors: list[dict], by_node: dict[str, list[Interval]],
                 node_ids: list[str]) -> dict[int, list[Interval]]:
    # ABSENCE OF A PROBE RECORD IS NOT EVIDENCE OF A BLACKOUT. An org with no
    # `node_alerts` rows at all gets no blind windows, and an org with no
    # registered probe gets none either — inventing a fleet-wide blackout out
    # of a missing row would be a fabrication in the other direction, and the
    # recording floors already cover "before the record can answer".
    shared: list[Interval] | None = None
    if node_ids:
        shared = by_node.get(node_ids[0], [])
        for n in node_ids[1:]:
            if not shared:
                break
            shared = intersect(shared, by_node.get(n, []))
    out: dict[int, list[Interval]] = {}
    for f in floors:
        node = f.get("assigned_node_id")
        ivs = by_node.get(node, []) if node else (shared or [])
        if ivs:
            out[f["device_id"]] = ivs
    return out


def replay(h, qs):
    # Reader-level and WORKER-REACHABLE, narrowed by the same
    # `visible_device_ids` helper every other list endpoint uses: a worker
    # replays exactly the fleet they can see live, and all three lists narrow
    # TOGETHER off one scope read (a span for a device whose floor was
    # withheld would render as a down mark on a device that is not on this
    # viewer's map).
    user = reader_or_401(h)
    if not user:
        return
    org = org_or_400(h, user, qs)
    if not org:
        return
    days = min(max(1, q_int_or(qs, "days", DEFAULT_DAYS)), MAX_DAYS)
    since_iso, until_iso = central_analytics.window(days)
    since, until = epoch_s(since_iso), epoch_s(until_iso)
    scope = visible_device_ids(h, user, org)

    floors = [f for f in h.store.replay_device_floors(org)
              if scope is None or f["device_id"] in scope]
    marks = h.store.node_stale_marks(org, since_iso)
    blind = device_blind(
        floors,
        stale_intervals(marks, since, until, h.cfg.central_node_stale_s),
        h.store.org_node_ids(org))

    spans = []
    for o in h.store.outages_in_window(org, since_iso, until_iso):
        did = int(o["device_id"])
        if scope is not None and did not in scope:
            continue
        # The TRUE start ships, never one clipped to the window: an outage
        # that began before `since` is still covering the whole left edge, and
        # a clipped start would read as "it dropped exactly when this window
        # opened". The client clips for drawing.
        spans.append({"outage_id": int(o["id"]), "device_id": did,
                      "start": epoch_s(o["started_at"]),
                      "end": epoch_s(o["resolved_at"]) if o["resolved_at"] else None,
                      "state": o["final_state"]})

    org_floor = h.store.org_recording_floor(org)
    h._reply(200, {
        "since": since, "until": until, "days": days,
        "now": epoch_s(until_iso),
        "org_since": epoch_s(org_floor) if org_floor else None,
        "devices": [{"device_id": f["device_id"],
                     "since": epoch_s(f["created_at"]) if f["created_at"] else None}
                    for f in floors],
        "spans": spans,
        "blind": [{"device_id": did, "start": s, "end": e}
                  for did, ivs in sorted(blind.items()) for s, e in ivs],
    })

File: web/src/lib/replay.ts
// THE RECONSTRUCTION — the one place this product decides what the record can
// say about a moment in the past. A pure sorted-interval walk: no React, no
// fetch, no clock. BOTH projections (the map replay and the outage Marey) read
// this and nothing else, so they can never disagree about when something
// dropped — the count-agreement rule applied to a timeline.
//
// THREE STATES, and `unknown` is the load-bearing one. The frozen doctrine
// says a reading nobody can stand behind must not render like a live one; the
// same rule pointed at the past says a MOMENT nobody was watching must not
// render like a moment that was fine. Painting a green pin is the worst lie
// this map can tell, so `unknown` covers every case where the record cannot
// answer:
//
//   1. PRE-HISTORY — before the org's own floor (`org_since`: the earlier of
//      when the org row appeared and its first outage). Nothing before that
//      was being recorded, so nothing before it can be claimed.
//   2. THE ENTITY DID NOT EXIST — before its own floor (`since`, a device's
//      created_at). A box entered on Tuesday was not "up" on Monday; it was
//      not there. Without this, a replay of last month renders today's fleet.
//   3. THE PROBE WAS SILENT — a `blind` window. The edge dials central, so a
//      dead probe produces no samples and the FSM freezes: the absence of an
//      outage row across those hours is not evidence of health.
//
// PRECEDENCE: down > unknown > up.
//   - `down` outranks `blind` because an outage span is a POSITIVE record
//     about that entity — the FSM opened it and only closed it when recovery
//     was observed — so a probe blackout inside an outage does not make the
//     outage unknown; it makes the recovery time uncertain, which the span's
//     own end already reports.
//   - `blind` outranks `up` because `up` is the only state here that is
//     inferred from ABSENCE, and absence is exactly what a blackout produces.
//   - a floor outranks everything: an entity that did not exist cannot have a
//     span covering it anyway, so this only ever guards nonsense.
//
// INTERVALS ARE HALF-OPEN, [start, end): an outage 10:00 to 11:00 is down at
// 10:00 and up at 11:00 — the same convention `day_availability` counts
// seconds under, so a Marey bar and the availability strip measure one thing.
//
// GENERIC BY ENTITY ID on purpose. Wave 3 feeds ONU transitions through the
// same walk with string ids (a MAC) and gets step-to-next-event, the Marey and
// the map projection for free; nothing here knows what a device is.

export type ReplayState = "up" | "down" | "unknown"
export type EntityId = number | string

export interface Span {
  start: number             // epoch seconds
  end: number | null        // null = still open at the end of the window
  // `false` marks a span that is a RESTATEMENT of somebody else's outage
  // (UNREACHABLE: my parent was down). It still renders as down-family — the
  // live map does — but it must never be COUNTED as this entity's downtime,
  // or one OLT's outage lands on every box behind it. Exactly the rule
  // `analytics.device_reliability` keeps by counting final_state == DOWN only.
  own?: boolean
}

export interface EntityInput {
  id: EntityId
  since?: number | null     // this entity's own recording floor
  down: Span[]
  blind?: Span[]
}

export interface ReplayWindow {
  since: number
  until: number
  now: number
  orgSince?: number | null
}

interface Norm {
  id: EntityId
  floor: number             // -Infinity = no floor known
  down: Array<[number, number]>
  owned: Array<[number, number]>
  blind: Array<[number, number]>
}

export interface Reconstruction {
  since: number
  until: number
  now: number
  ids: EntityId[]
  /** Every transition inside the window, sorted and de-duplicated. */
  events: number[]
  stateAt(id: EntityId, t: number): ReplayState
  statesAt(t: number): Map<EntityId, ReplayState>
  nextEventAfter(t: number): number | null
  prevEventBefore(t: number): number | null
  /**
   * The last transition at or before `t`, falling back to the window start.
   * BETWEEN TWO EVENTS NOTHING CHANGES — that is what an event is — so this
   * is the cheapest possible key for anything that renders a whole fleet at
   * T: it moves only when the picture does. Playback runs at ten frames a
   * second, and without it every frame would rebuild the device rows, the
   * clusters and the link set to draw an identical map.
   */
  eventFloorAt(t: number): number
  /** Merged down intervals, clipped to the window (for drawing bars). */
  downBars(id: EntityId): Array<[number, number]>
  /** Merged unknown intervals, clipped to the window (floor included). */
  unknownBars(id: EntityId): Array<[number, number]>
  /** Seconds this entity was down BY ITS OWN FAULT inside [from, to). */
  downSecondsIn(id: EntityId, from: number, to: number): number
  outageCountIn(id: EntityId, from: number, to: number): number
}

const OPEN = Number.POSITIVE_INFINITY

function merge(spans: Span[], keep: (s: Span) => boolean): Array<[number, number]> {
  const raw: Array<[number, number]> = []
  for (const s of spans) {
    if (!keep(s)) continue
    const end = s.end == null ? OPEN : s.end
    if (end > s.start) raw.push([s.start, end])
  }
  raw.sort((a, b) => a[0] - b[0] || a[1] - b[1])
  const out: Array<[number, number]> = []
  for (const iv of raw) {
    const last = out[out.length - 1]
    // Adjacent spans (one ends exactly where the next begins) merge: a flap
    // that re-opened in the same second is one bar, not two hairlines.
    if (last && iv[0] <= last[1]) last[1] = Math.max(last[1], iv[1])
    else out.push([iv[0], iv[1]])
  }
  return out
}

function covers(ivs: Array<[number, number]>, t: number): boolean {
  // Binary search: the map calls this once per device per scrub tick, and a
  // fleet-scale linear scan through a 90-day span list is the one place this
  // walk could get slow.
  let lo = 0
  let hi = ivs.length - 1
  while (lo <= hi) {
    const mid = (lo + hi) >> 1
    if (t < ivs[mid][0]) hi = mid - 1
    else if (t >= ivs[mid][1]) lo = mid + 1
    else return true
  }
  return false
}

function clip(ivs: Array<[number, number]>, since: number, until: number
): Array<[number, number]> {
  const out: Array<[number, number]> = []
  for (const [a, b] of ivs) {
    const lo = Math.max(a, since)
    const hi = Math.min(b, until)
    if (hi > lo) out.push([lo, hi])
  }
  return out
}

function overlapSeconds(ivs: Array<[number, number]>, from: number, to: number): number {
  let total = 0
  for (const [a, b] of ivs) {
    const lo = Math.max(a, from)
    const hi = Math.min(b, to)
    if (hi > lo) total += hi - lo
  }
  return total
}

export function buildReconstruction(
  entities: EntityInput[], win: ReplayWindow,
): Reconstruction {
  const { since, until, now } = win
  const orgFloor = win.orgSince ?? Number.NEGATIVE_INFINITY
  const byId = new Map<EntityId, Norm>()
  const eventSet = new Set<number>()

  const mark = (t: number) => {
    if (t > since && t < until) eventSet.add(t)
  }

  for (const e of entities) {
    const floor = Math.max(orgFloor, e.since ?? Number.NEGATIVE_INFINITY)
    const down = merge(e.down, () => true)
    const owned = merge(e.down, (s) => s.own !== false)
    const blind = merge(e.blind ?? [], () => true)
    byId.set(e.id, { id: e.id, floor, down, owned, blind })
    if (Number.isFinite(floor)) mark(floor)
    for (const ivs of [down, blind]) {
      for (const [a, b] of ivs) { mark(a); if (b !== OPEN) mark(b) }
    }
  }

  const events = [...eventSet].sort((a, b) => a - b)
  const ids = entities.map((e) => e.id)

  const stateAt = (id: EntityId, t: number): ReplayState => {
    const n = byId.get(id)
    // An entity the reply never mentioned is unanswerable, not healthy — a
    // device created after this window was fetched lands here.
    if (!n) return "unknown"
    if (t < n.floor) return "unknown"
    if (covers(n.down, t)) return "down"
    if (covers(n.blind, t)) return "unknown"
    return "up"
  }

  const unknownBars = (id: EntityId): Array<[number, number]> => {
    const n = byId.get(id)
    if (!n) return [[since, until]]
    const bars: Array<[number, number]> = []
    if (n.floor > since) bars.push([since, Math.min(n.floor, until)])
    for (const b of clip(n.blind, since, until)) bars.push(b)
    return merge(bars.map(([a, b]) => ({ start: a, end: b })), () => true)
  }

  return {
    since, until, now, ids, events, stateAt,
    statesAt: (t) => {
      const out = new Map<EntityId, ReplayState>()
      for (const id of ids) out.set(id, stateAt(id, t))
      return out
    },
    nextEventAfter: (t) => {
      for (const e of events) if (e > t) return e
      return null
    },
    prevEventBefore: (t) => {
      for (let i = events.length - 1; i >= 0; i--) if (events[i] < t) return events[i]
      return null
    },
    eventFloorAt: (t) => {
      let lo = 0
      let hi = events.length - 1
      let best = since
      while (lo <= hi) {
        const mid = (lo + hi) >> 1
        if (events[mid] <= t) { best = events[mid]; lo = mid + 1 } else hi = mid - 1
      }
      return best
    },
    downBars: (id) => clip(byId.get(id)?.down ?? [], since, until),
    unknownBars,
    downSecondsIn: (id, from, to) =>
      overlapSeconds(byId.get(id)?.owned ?? [], from, to),
    outageCountIn: (id, from, to) =>
      (byId.get(id)?.owned ?? []).filter(([a, b]) => b > from && a < to).length,
  }
}
